backend/src/routes/api/notebooks/index.ts

```diff
@@ -61,39 +63,64 @@ module.exports = async (fastify: KubeFastifyInstance) => {
   );
 
   fastify.post(
-    '/:namespace',
+    '/',
     secureRoute(fastify)(
       async (
         request: FastifyRequest<{
-          Params: {
-            namespace: string;
-          };
-          Body: Notebook;
+          Body: NotebookData;
         }>,
+        reply: FastifyReply,
       ) => {
-        return createNotebook(fastify, request);
+        const notebookData = request.body;
+        const username = await getUserName(fastify, request);
+
+        return createNotebook(fastify, notebookData, username).catch((e) => {
+          fastify.log.error(
+            `Failed to create notebook, ${e.response?.data?.message || e.message}}`,
+          );
+          reply.status(400).send(e.response?.data?.message || e.message);
+        });
+      },
+    ),
+  );
+
+  fastify.put(
+    '/',
+    secureRoute(fastify)(
+      async (
+        request: FastifyRequest<{
+          Body: NotebookData;
+        }>,
+        reply: FastifyReply,
+      ) => {
+        const notebookData = request.body;
+        const username = await getUserName(fastify, request);
+
+        return updateNotebook(fastify, notebookData, username).catch((e) => {
+          fastify.log.error(
+            `Failed to update notebook, ${e.response?.data?.message || e.message}}`,
+          );
+          reply.status(400).send(e.response?.data?.message || e.message);
+        });
       },
     ),
   );
 
-  fastify.patch(
+  fastify.delete(
```

Comment:
No delete notebook endpoint please... this is not a needed feature.

Reply:
mmmm ok, what do you suggest, add a value for the PUT method and and if it's enable stop the notebook, or just a straight up endpoint to just stop. I would rather have the stop functionality separated somewhere else.

Reply:
I listed endpoints what I thought were good here based on what the UI needs to do -- https://github.com/opendatahub-io/odh-dashboard/issues/581

Thinking a bit more on it -- I think essentially we just need:
1. `GET` Notebook (fairly straight forward)
1. `GET` Notebook status (gets the notebook + its running status, which is more than just 
1. `POST` "Create" Notebook (which takes form data instead of a CR)
    - `PUT` "Restart" Notebook (which takes form data instead of a CR) -- honestly this could probably be part of the `POST` call and just check if we have a notebook, update/restart it, if we don't create it
1. `POST` "Stop" Notebook (takes nothing, applies the annotation server side)

4 endpoints might be enough to do all we need and nothing takes a CR anymore.

Reply:
Ok, seems great, sorry i get carried away and i forgot you had the spec, i'm gonna uodate the pr with this, it wont take much time, thanks @andrewballantyne

Reply:
Ok, updated with this endpoints you suggested, I've also updated the type `NotebookData`, planning to remove the following fields too:

* `imageSelection`& `imageUrl`: Just have an `image` field and then get the values in the backend, this will prevent any injection of undesired images.
*`envVarFileName` and just generate it in the backend, avoid sanitization for the name, add similar checks for the secrets and configmap values just to avoid the use of undesired values.

Apart from that, I've tested that we can move the PVC creation and the env creation to the same request of stop/restart, I wanna know your thoughts on that @cfchase @andrewballantyne.

Reply:
Please do me a favor and don't put verbs like create/stop/start verbs in restful design.  It's an antipattern that absolutely drives me nuts.  the verbs should be the HTTP verbs.

https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design#organize-the-api-design-around-resources

Reply:
Ok fair, I followed this patter first but then I changed it, seems weird the stop status, I went first with `DELETE` since we were removing the resource of the notebook, I'm gonna put the update and stop inside the `PUT` and then the creation in `POST` to follow the specification. Thanks Chris.

Reply:
a PUT is generally a full replacement of a resource.  PATCH for partial.

Reply:
POST for stop was my suggestion -- I guess I have seen a lot of endpoints in various stages of bastardization of REST... POST seems to be the "do action" in most APIs I have seen. PATCH was intended for providing some indication of what to adjust -- whereas here we would not ask anything of the client, a dedicated endpoint to do something `POST /api/notebooks/stop` seemed very reasonable to me. I of course defer to Chris, I'm an outsider when it comes to backend design.

Reply:
POST has been the catchall abused verb since the beginning, but I'd rather not resort to that unless you really can't express the intent some better way.

Reply:
I agree Chris -- we probably shouldn't favour POST for the catchall situations and I'll bow to your judgement here. 

However PATCH seems wrong to me on the principle of "stop has nothing to do with PATCH" as far as our REST API is concerned. It's just how we talk to k8s to achieve the kubeflow desire of patching an annotation. Seems "leaking implementation" to me... not that doing that is a problem, just not one I agree with fully.

Anyways, I should stay out of backend design 🙂 PATCH for stop will work fine from an implementation point of view. 👍

Reply:
The main problem here is that following the full specification both of you are right:

* `PUT` should be idempotent and replace the hole object. (We are trying to avoid this since we don’t wanna mess around with the hole notebook spect).
* `PATCH` should use a patch language like `JSON` or `XML`, the body itself should contain the instructions to update the object but it doesn’t need to be idempotent.

Trying to hide away the spec of notebook makes both verbs harder to follow, so right now we can follow the specification we agree on (trying to avoid antipatterns like the one chris pointed out).

I’m fine either having stop as a `PUT` or `PATCH`, whatever you guys think it’s the best solution.

Reply:
If it’s a `PATCH`, we can add a sanitization in the backend to ensure the only JSON permitted is:

`metadata: { annotations: { 'kubeflow-resource-stopped': <date>} } `

With that, we ensure are following the `PATCH` spec and we are avoiding the modification of the rest of the notebook object.

What do you think?

Reply:
I like it.

Reply:
Eliminate all you can from the frontend -- the frontend can talk "actions" not data... if we do that, then there can't be security holes, because they talk actions and we do all the data in a secure location.

Ideally, the call to stop a notebook requires NOTHING from the caller. It is a direct endpoint to say "do action" and thus however we need to do it can be done by us. If the caller needs to provide direction, let us make that as small as possible -- like if they need to provide a date, they JUST provide a date, we do the annotation and formatting of the date on the backend. 

The client can't be trusted, so make it hard for them to "lie" about what they are doing.

Reply:
The more the client gives us, the more sanitization we need to do... let us create as little of those points as possible to avoid any chance the UI can sneak something in.

Reply:
Yes, you are right too @andrewballantyne , I mean, to be fair, the one verb that fits the best action to stop for me is `DELETE` 😂. At the end of the day we are deleting the notebook, although we are just patching the spec in the backend. I agree adding PATCH adds more vulnerable surface. We are having a great discussion here, I would gladly accept both opinions.

Reply:
I retract any statement of verbs I have made -- just make it so the client only provides the extreme bare min -- easier for us to validate. Honestly, the backend can stop the Notebook at it's own timestamp if it wants to. Less chance the client can try and fudge the data.

Reply:
we are not deleting the notebook resource

Reply:
@cfchase If you infer the date from the server and not the client -- yeah, I think stop is as simple as `PATCH /api/notebooks` with `{ state: 'stopped' }` and let the server apply the annotation & figure out the current datestamp to use as the stop.

Reply:
I'm implementing the suggestion you are discussing above as I think is the best way to do it too. Refactoring.